src/lib/shadow-shim-helper-rs/src/scmutex.rs

```diff
@@ -145,61 +292,75 @@ unsafe impl<T> VirtualAddressSpaceIndependent for SelfContainedMutex<T> where
 {
 }
 
-const UNLOCKED: i32 = 0;
-const LOCKED: i32 = 1;
-const LOCKED_DISCONNECTED: i32 = 2;
+const UNLOCKED: u16 = 0;
+const LOCKED: u16 = 1;
+const LOCKED_DISCONNECTED: u16 = 2;
 
 impl<T> SelfContainedMutex<T> {
     pub fn new(val: T) -> Self {
         Self {
-            futex: sync::AtomicI32::new(UNLOCKED),
-            sleepers: sync::AtomicU32::new(0),
+            futex: AtomicFutexWord::new(FutexWord {
+                lock_state: UNLOCKED,
+                num_sleepers: 0,
+            }),
             val: sync::UnsafeCell::new(val),
         }
     }
 
     pub fn lock(&self) -> SelfContainedMutexGuard<T> {
+        // On first attempt, optimistically assume the lock is uncontended.
+        let mut current = FutexWord {
+            lock_state: UNLOCKED,
+            num_sleepers: 0,
+        };
         loop {
-            // Try to take the lock.
-            //
-            // We use `Acquire` on the failure path as well to ensure that the
-            // `sleepers.fetch_add` is observed strictly after this operation.
-            let prev = self.futex.compare_exchange(
-                UNLOCKED,
-                LOCKED,
-                sync::Ordering::Acquire,
-                sync::Ordering::Acquire,
-            );
-            let prev = match prev {
-                Ok(_) => {
-                    // We successfully took the lock.
-                    break;
+            if current.lock_state == UNLOCKED {
+                // Try to take the lock.
+                //
+                // I *think* we could technically just do a 16-bit atomic
+                // operation on the lock-state part of the futex word, but that
+                // is incompatible with validating via loom, and more fragile in
+                // general.
+                let current_res = self.futex.compare_exchange(
+                    current.into(),
+                    FutexWord {
+                        lock_state: LOCKED,
+                        num_sleepers: current.num_sleepers,
+                    }
+                    .into(),
+                    sync::Ordering::Acquire,
+                    sync::Ordering::Relaxed,
+                );
+                current = match current_res {
+                    Ok(_) => {
+                        // We successfully took the lock.
+                        break;
+                    }
+                    // We weren't able to take the lock.
+                    Err(i) => i,
+                };
+            }
+
+            // If the lock is available, try again now that we've sync'd the
+            // rest of the futex word (num_sleepers).
+            if current.lock_state == UNLOCKED {
+                continue;
+            }
+
+            // Update the sleeper count.
+            current = self.futex.inc_sleepers_and_fetch(sync::Ordering::Relaxed);
```

Comment:
Something I'm not clear about: Does a relaxed-store break the release-acquire ordering? For example:

```
Thread A:
mutex.unlock()
  1. uses a release-store so that memory before this store will be seen after a
     later acquire-load

Thread B:
mutex.lock()
  1. the `compare_exchange` fails
  2. we do a `inc_sleepers_and_fetch(Relaxed)`, which is a relaxed-store
  3. the mutex became unlocked, so we do a `dec_sleepers_and_fetch(Relaxed)`,
     which is also a relaxed-store
  4. in the second loop iteration, the `compare_exchange` is successful, which is
     an acquire-load
```

Do these two relaxed-store operations in thread B break the release-acquire ordering for the acquire-load in step 4? Or will the release-acquire ordering still be guaranteed, even though there were relaxed stores in between?

On the other hand, what if these 2 relaxed-stores were changed to release-stores? Would the release-acquire ordering then only be guaranteed between this last release-store (which would be step 3) and this aquire-load in step 4?

The [question](https://stackoverflow.com/questions/71509935/how-does-mixing-relaxed-and-acquire-release-accesses-on-the-same-atomic-variable) seems to be related, but I haven't fully digested it yet.

Reply:
I don't *think* the relaxed operations occurring between the acquire and release interfere with those.

e.g. from the [rustonomicon](https://doc.rust-lang.org/nomicon/atomics.html#acquire-release) "Every write (including non-atomic and relaxed atomic writes) that happened before A's release will be observed by B after its acquisition."

Looking at this part again I did start to wonder if the relaxed increment of the sleeper count might not be observed by another thread that's unlocking, causing us to lose a wakeup. I think it doesn't currently happen in our loom model due to the mutex used to implement the global futex set. i.e. this could be a bug in our loom model hiding a bug in the lock itself, if our model of the futex syscall provides stronger synchronization than a real futex syscall.

I don't think that it does, though. My reading of the futex man page makes me think that the futex syscall is effectively CST, and so really does synchronize the increment for us.

> The loading of the futex word's value, the comparison of that value with the expected value, and the actual blocking will happen atomically and will be totally ordered with respect to concurrent operations performed by other threads  on
> the same futex word.  Thus, the futex word is used to connect the synchronization in user space with the implementation of blocking by the kernel.  Analogously to an atomic compare-and-exchange operation that potentially changes shared memory,
> blocking via a futex is an atomic compare-and-block operation.

I *think* the Acquire-Release ordering we get from locking and unlocking in our model of futex is a little weaker than the real guarantee, which is ok; there's only a problem if it's stronger than the real guarantee.

It might be interesting to see if we could implement a futex model with weaker synchronization, to be more confident we're not relying on stronger synchronization than the syscall actually gives us. I think to do so would end up requiring basically a spin loop though; even the lower level [Notify](https://docs.rs/loom/0.5.6/loom/sync/struct.Notify.html) primitive "establishes an aquire / release synchronization point".

I'm inclined to leave as-is for now, but maybe try to capture some of this in comments. Wdyt?

Reply:
> On the other hand, what if these 2 relaxed-stores were changed to release-stores? Would the release-acquire ordering then only be guaranteed between this last release-store (which would be step 3) and this aquire-load in step 4?

I'm not sure I completely grok this part of your question, but in general I don't think extra atomic operations would break the overall ordering guarantees.

My understanding is that for an acquire-release pair, the guarantee is that if T2 observes a value that was stored via Release ordering by T1, using Acquire ordering, then it is guaranteed to also observe all of T1's memory stores that it did before that Release store.

I think we could run into the sort of problems you're thinking about if we ever did a relaxed store setting the state to UNLOCKED. Even if we did it just before a stronger store setting to UNLOCKED, a thread loading the value and seeing UNLOCKED wouldn't have any ordering guarantees, since it could be observing the relaxed store.

Reply:
> I don't think the relaxed operations occurring between the acquire and release interfere with those.
>
> e.g. from the rustonomicon "Every write (including non-atomic and relaxed atomic writes) that happened before A's release will be observed by B after its acquisition."

The line that makes me question this is:

https://en.cppreference.com/w/cpp/atomic/memory_order

> If an atomic store in thread A is a release operation, an atomic load in thread B from the same variable is an acquire operation, **and the load in thread B reads a value written by the store in thread A**, then the store in thread A synchronizes-with the load in thread B.

By "thread B reads a value written by the store in thread A", does it mean the specific value written during the store? What if there are other relaxed stores between them that change the value?

There is also:

> That is, once the atomic load is completed, thread B is guaranteed to see everything thread A wrote to memory. **This promise only holds if B actually returns the value that A stored, or a value from later in the release sequence.**

So it sounds like it must load the same value as stored by the original require-store, or by another store in the "release sequence".

It then defines "release sequence" as:

> After a release operation A is performed on an atomic object M, the longest continuous subsequence of the modification order of M that consists of
> 1) Writes performed by the same thread that performed A (until C++20)
> 2) Atomic read-modify-write operations made to M by any thread
>
> is known as release sequence headed by A 

I'm not sure what "until C++20" means here (maybe this line is no longer true?). But it seems to me like if another thread does a relaxed-store, then it's definitely not part of the release sequence and breaks the release-acquire ordering. If the same thread does a relaxed-store, then that depends on if line (1) holds or not for us in rust.

> My understanding is that for an acquire-release pair, the guarantee is that if T2 observes a value that was stored via Release ordering by T1, using Acquire ordering, then it is guaranteed to also observe all of T1's memory stores that it did before that Release store.

I think this is true, but that the acquire-load doesn't pair with all past release-stores, only any release-stores in the current release sequence. And that release sequence can be broken by other non-release stores or non-acquire loads, which breaks the guarantee.

> I'm inclined to leave as-is for now, but maybe try to capture some of this in comments. Wdyt?

Yeah I think it's definitely an improvement, so I'm good with merging. I don't understand the memory ordering stuff well enough to know if everything is correct, but we can always revisit again later if we learn more.

Reply:
Actually I'm not sure if/why the mutex inside the futex model helps us after all. A thread going to sleep on the futex will have effectively done a synchronized *store* of the sleeper count because of the mutex, but a thread doing the Release compare-and-swap to release the lock still shouldn't be guaranteed to see that incremented sleeper count, since it happens before doing a futex wake operation.

I think we'd be justified in making our model stronger by using strong ordering and/or a fence inside the futex_wait model, based on the futex docs, but I'd like to understand why loom doesn't find a problem in the current model.

Setting aside now and hopefully for the break, but will revisit :)

Reply:
> By "thread B reads a value written by the store in thread A", does it mean the specific value written during the store? What if there are other relaxed stores between them that change the value?

I think/hope this is short-hand for something more like "thread B reads a value that implies we're seeing the store in thread A", but agree it'd be good to find something in the spec to validate that interpretation...

I made a quick test program to validate that loom agrees with this interpretation, but maybe it'd be good to check with the loom authors and/or find better support in the spec that this isn't a deficiency in loom:

```
use loom::cell::UnsafeCell;
use loom::sync::Arc;
use loom::sync::atomic::{AtomicU32, Ordering};
use loom::thread;

pub fn main() {
    loom::model(|| {
        let x = Arc::new(UnsafeCell::new(0u32));
        let a = Arc::new(AtomicU32::new(0));

        let writer = {
            let x = x.clone();
            let a = a.clone();
            thread::spawn(move || {
                unsafe { *x.get_mut().deref() = 1 };
                a.store(1, Ordering::Release);
                // Does this break the Release-Acquire guarantee?
                a.fetch_add(1, Ordering::Relaxed);
            })
        };

        // Note that we're also accepting 2 here, which isn't the precise value written by the
        // Release operation, but *implies* that we're observing the Release operation.
        if a.load(Ordering::Acquire) > 0 {
            let x: u32 = unsafe { *x.get().deref() };
            assert!(x == 1);
        }

        writer.join().unwrap();
    });
}

```

Reply:
> This promise only holds if B actually returns the value that A stored, or a value from later in the release sequence.

> After a release operation A is performed on an atomic object M, the longest continuous subsequence of the modification order of M that consists of
> 1) Writes performed by the same thread that performed A (until C++20)
> 2) Atomic read-modify-write operations made to M by any thread
>
> is known as release sequence headed by A 

I think a key point that may have been missed above is that the relaxed increments and decrements are read-modify-write operations; not just stores. Relaxed RMW operations are still part of the release sequence.

e.g. this is similar to figure 5 in the [paper loom is based on](http://demsky.eecs.uci.edu/publications/c11modelcheck.pdf):

![image](https://user-images.githubusercontent.com/179146/204025121-55bccf10-7148-491b-8762-ac84f347f330.png)

So I'm feeling more confident now that this is correct, but yeah still good to capture some of this conversation in comments.

Reply:
Added comments at the increment and decrement operations

Reply:
Ah, nice diagram and paper!

> I think a key point that may have been missed above is that the relaxed increments and decrements are read-modify-write operations; not just stores. Relaxed RMW operations are still part of the release sequence.

Okay yeah that sounds right. So (B) in the diagram is part of the release sequence since it's a "write performed by the same thread that performed" (A), and (C) and (D) are part of the release sequence since they're "atomic read-modify-write operations made to [v] by any thread", even though they're only relaxed and non release-acquire.